File: src/ClientModel/PrimitiveType.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

using System;
using System.Collections.Generic;

namespace AutoRest.Java.Model
{
    /// <summary>
    /// A basic type used by a REST API method.
    /// </summary>
    public class PrimitiveType : IType
    {
        public static readonly PrimitiveType Void = new PrimitiveType("void", ClassType.Void);
        public static readonly PrimitiveType Boolean = new PrimitiveType("boolean", ClassType.Boolean, (string defaultValueExpression) => defaultValueExpression.ToLowerInvariant());
        public static readonly PrimitiveType Byte = new PrimitiveType("byte", ClassType.Byte);
        public static readonly PrimitiveType Int = new PrimitiveType("int", ClassType.Integer);
        public static readonly PrimitiveType Long = new PrimitiveType("long", ClassType.Long, (string defaultValueExpression) => defaultValueExpression + 'L');
        public static readonly PrimitiveType Double = new PrimitiveType("double", ClassType.Double, (string defaultValueExpression) => double.Parse(defaultValueExpression).ToString());

        public static readonly PrimitiveType UnixTimeLong = new PrimitiveType("long", ClassType.UnixTimeLong);

        /// <summary>
        /// Create a new PrimitiveType from the provided properties.
        /// </summary>
        /// <param name="name">The name of this type.</param>
        private PrimitiveType(string name, ClassType nullableType, Func<string,string> defaultValueExpressionConverter = null)
        {
            Name = name;
            NullableType = nullableType;
            DefaultValueExpressionConverter = defaultValueExpressionConverter;
        }

        /// <summary>
        /// The name of this type.
        /// </summary>
        public string Name { get; }

        /// <summary>
        /// The nullable version of this primitive type.
        /// </summary>
        private ClassType NullableType { get; }

        public void AddImportsTo(ISet<string> imports, bool includeImplementationImports)
        {
        }

        public IType AsNullable()
        {
            return NullableType;
        }

        public bool Contains(IType type)
        {
            return this == type;
        }

        private Func<string,string> DefaultValueExpressionConverter { get; }

        public string DefaultValueExpression(string sourceExpression)
        {
            string result = sourceExpression;
            if (result != null && DefaultValueExpressionConverter != null)
            {
                result = DefaultValueExpressionConverter(sourceExpression);
            }
            return result;
        }

        public IType ConvertToClientType()
        {
            IType clientType = this;
            if (this == PrimitiveType.UnixTimeLong)
            {
                clientType = ClassType.UnixTimeDateTime;
            }
            return clientType;
        }

        public override string ToString()
        {
            return Name;
        }
    }
}

File: mapper/src/main/java/com/azure/autorest/model/clientmodel/PrimitiveType.java
package com.azure.autorest.model.clientmodel;

import java.util.*;

// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.


/** 
 A basic type used by a client.
*/
public class PrimitiveType implements IType
{
    public static final PrimitiveType Void = new PrimitiveType("void", ClassType.Void);
    public static final PrimitiveType Boolean = new PrimitiveType("boolean", ClassType.Boolean, (String defaultValueExpression) -> defaultValueExpression.ToLowerInvariant());
//C# TO JAVA CONVERTER WARNING: Unsigned integer types have no direct equivalent in Java:
//ORIGINAL LINE: public static readonly PrimitiveType Byte = new PrimitiveType("byte", ClassType.Byte);
    public static final PrimitiveType Byte = new PrimitiveType("byte", ClassType.Byte);
    public static final PrimitiveType Int = new PrimitiveType("int", ClassType.Integer);
    public static final PrimitiveType Long = new PrimitiveType("long", ClassType.Long, (String defaultValueExpression) -> defaultValueExpression + 'L');
    public static final PrimitiveType Double = new PrimitiveType("double", ClassType.Double, (String defaultValueExpression) -> Double.parseDouble(defaultValueExpression).toString());

    public static final PrimitiveType UnixTimeLong = new PrimitiveType("long", ClassType.UnixTimeLong);

    /**
     Create a new PrimitiveType from the provided properties.

     @param name The name of this type.
    */

    private PrimitiveType(String name, ClassType nullableType)
    {
        this(name, nullableType, null);
    }

//C# TO JAVA CONVERTER NOTE: Java does not support optional parameters. Overloaded method(s) are created above:
//ORIGINAL LINE: private PrimitiveType(string name, ClassType nullableType, Func<string,string> defaultValueExpressionConverter = null)
    private PrimitiveType(String name, ClassType nullableType, java.util.function.Function<String,String> defaultValueExpressionConverter)
    {
        Name = name;
        NullableType = nullableType;
        DefaultValueExpressionConverter = (String arg) -> defaultValueExpressionConverter.invoke(arg);
    }

    /**
     The name of this type.
    */
    private String Name;
    public final String getName()
    {
        return Name;
    }

    /**
     The nullable version of this primitive type.
    */
    private ClassType NullableType;
    private ClassType getNullableType()
    {
        return NullableType;
    }

    public final void AddImportsTo(Set<String> imports, boolean includeImplementationImports)
    {
    }

    public final IType AsNullable()
    {
        return getNullableType();
    }

    public final boolean Contains(IType type)
    {
        return this == type;
    }

    private java.util.function.Function<String,String> DefaultValueExpressionConverter;
    private java.util.function.Function<String,String> getDefaultValueExpressionConverter()
    {
        return DefaultValueExpressionConverter;
    }

    public final String DefaultValueExpression(String sourceExpression)
    {
        String result = sourceExpression;
        if (result != null && getDefaultValueExpressionConverter() != null)
        {
            result = DefaultValueExpressionConverter(sourceExpression);
        }
        return result;
    }

    public final IType getClientType()
    {
        IType clientType = this;
        if (this == PrimitiveType.UnixTimeLong)
        {
            clientType = ClassType.UnixTimeDateTime;
        }
        return clientType;
    }

    public final String ConvertToClientType(String expression)
    {
        if (getClientType() == this)
        {
            return expression;
        }

        if (this == PrimitiveType.UnixTimeLong)
        {
            expression = String.format("OffsetDateTime.from(Instant.ofEpochSecond(%1$s));", expression);
        }
        return expression;
    }

    public final String ConvertFromClientType(String expression)
    {
        if (getClientType() == this)
        {
            return expression;
        }

        if (this == PrimitiveType.UnixTimeLong)
        {
            expression = String.format("%1$s.toEpochSecond()", expression);
        }
        return expression;
    }

    @Override
    public String toString()
    {
        return getName();
    }
}